### PR TITLE
OpenMPTarget: fixes and workarounds to work with "Release" build type.

### PR DESCRIPTION
**Pull request by rgayatri23**
Fixes
1. Changed the testing from "Debug" build type to "RelWithDebInfo" build type
2. team_size for scratch tests in incremental tests are set to multiples of 32 for the OPENMPTARGET backend to avoid hanging in the Release builds.
3. `omp_target_memcpy` calls are protected to be called only when the copy-size > 0 .
4.  Added a `nowait` clause to the `TeamVectorRange` implementation. Avoids hanging in Release builds and is also accurate for the corresponding Kokkos constructs.
5. Following unit tests fail in the Release build `*reducers_size_t:*reducers_double:*reducers_struct` . WIP to fix them.

### CODE REVIEW

File: core/src/Kokkos_OpenMPTargetSpace.hpp [hunk 1]
```diff
@@ -217,41 +217,47 @@ template <class ExecutionSpace>
 struct DeepCopy<Kokkos::Experimental::OpenMPTargetSpace,
                 Kokkos::Experimental::OpenMPTargetSpace, ExecutionSpace> {
   DeepCopy(void* dst, const void* src, size_t n) {
-    omp_target_memcpy(dst, const_cast<void*>(src), n, 0, 0,
-                      omp_get_default_device(), omp_get_default_device());
+    if (n)
```
Comment (dalg24) [inline core/src/Kokkos_OpenMPTargetSpace.hpp:220]: Readability
```suggestion
    if (n > 0)
```

Comment (dalg24) [reply core/src/Kokkos_OpenMPTargetSpace.hpp:220]: I'd like you to add a comment in the code where you elaborate on why that call to `omp_target_memcpy` is protected.  The specification does not explicitly calls out whether it is legal to pass null pointers with a zero length (see https://www.openmp.org/spec-html/5.1/openmpsu172.html).  Did you check whether this was the issue?

Also, any good reason to ignore the return value here?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 2]
```diff
@@ -593,7 +593,9 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
     void* scratch_ptr = OpenMPTargetExec::get_scratch_ptr();
     FunctorType a_functor(m_functor);
 
-    // Maximum active teams possible.
+    // FIXME_OPENMPTARGET - If the team_size is not a multiple of 32, the
+    // scratch implementation does not work in the Release or RelWithDebugInfo
+    // mode but works in the Debug mode. Maximum active teams possible.
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:598]:
```suggestion
    // FIXME_OPENMPTARGET - If the team_size is not a multiple of 32, the
    // scratch implementation does not work in the Release or RelWithDebugInfo
    // mode but works in the Debug mode.
    // Maximum active teams possible.
```

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_OpenMPTargetSpace.hpp (hunk 1 above)
Comment (rgayatri23) [reply core/src/Kokkos_OpenMPTargetSpace.hpp]: Yes the specification does not say anything specific about passing zero length to `omp_target_memcpy`. I guess the Release build just protects against unnecessary copies. 
I added a return value to the memcpy calls and am checking for its success.

**Comment by masterleinad** [comment]
The failing SYCL build is fixed by #3741.

**Review by dalg24 — APPROVED** [review]

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp [hunk 3]
```diff
@@ -120,9 +120,15 @@ struct ThreadScratch {
 
 TEST(TEST_CATEGORY, IncrTest_12a_ThreadScratch) {
   ThreadScratch<TEST_EXECSPACE> test;
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+  test.run(1, 32, 9);
+  test.run(2, 64, 22);
+  test.run(14, 128, 321);
+#else
```
Comment (masterleinad) [inline core/unit_test/incremental/Test12a_ThreadScratch.hpp:127]: What is the reason for changing this?

File: core/unit_test/incremental/Test12b_TeamScratch.hpp [hunk 4]
```diff
@@ -107,11 +107,13 @@ struct TeamScratch {
 
 TEST(TEST_CATEGORY, IncrTest_12b_TeamScratch) {
   TeamScratch<TEST_EXECSPACE> test;
-  test.run(1, 4, 4);
-  test.run(4, 7, 10);
 #ifdef KOKKOS_ENABLE_OPENMPTARGET
-  test.run(14, 277, 1);
+  test.run(1, 32, 4);
+  test.run(4, 64, 10);
+  test.run(14, 128, 20);
```
Comment (masterleinad) [inline core/unit_test/incremental/Test12b_TeamScratch.hpp:113]: and this?

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 3 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:127]: In the Release and RelWithDebInfo builds, the scratch implementation fails unless the team_size is a multiple of 32.

File: core/unit_test/incremental/Test12b_TeamScratch.hpp (hunk 4 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test12b_TeamScratch.hpp:113]: Same as above, the team_size needs to be a multiple of 32 for Release builds.

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 3 above)
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:127]: Would you mind adding a FIXME with this comment (also in the other place)?

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Comment by masterleinad** [comment]
Retest this please.

File: core/unit_test/TestReducers_d.hpp [hunk 5]
```diff
@@ -54,8 +54,15 @@ TEST(TEST_CATEGORY, reducers_complex_double) {
 TEST(TEST_CATEGORY, reducers_struct) {
   TestReducers<array_reduce<float, 1>, TEST_EXECSPACE>::test_sum(1031);
   TestReducers<array_reduce<float, 2>, TEST_EXECSPACE>::test_sum(1031);
-  TestReducers<array_reduce<float, 3>, TEST_EXECSPACE>::test_sum(1031);
   TestReducers<array_reduce<float, 4>, TEST_EXECSPACE>::test_sum(1031);
+  // FIXME_OPENMPTARGET - The size of data in array_reduce has to be a power of
+  // 2 for OPENMPTARGET backend in Release and RelWithDebInfo builds.
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+  TestReducers<array_reduce<float, 4>, TEST_EXECSPACE>::test_sum(1031);
```
Comment (crtrott) [inline core/unit_test/TestReducers_d.hpp:61]: this guy is now twice here isn't it?

**Review by crtrott — DISMISSED** [review]
one of the test lines is duplicated see my comment.

Otherwise fine.

**Comment by rgayatri23** [comment]
> one of the test lines is duplicated see my comment.
> 
> Otherwise fine.

Fixed it.

**Pull request merged**